Lib/idlelib/idle_test/test_sidebar.py

```diff
@@ -370,5 +391,326 @@ def assert_colors_are_equal(colors):
+        return [canvas.bbox(text)[1] for text in texts]
+
+    def assert_sidebar_lines_synced(self):
+        self.assertEqual(
+            self.get_sidebar_line_y_coords(),
+            self.get_shell_line_y_coords(),
+        )
+
+    def do_input(self, input):
+        shell = self.shell
+        text = shell.text
+        for line_index, line in enumerate(input.split('\n')):
+            if line_index > 0:
+                text.event_generate('<<newline-and-indent>>')
+            for char in line:
+                text.insert('insert', char)
+
+    def run_test_coroutine(self, coroutine):
+        root = self.root
+        # Exceptions raised by self.assert...() need to be raised outside of
+        # the after callback in order for the test harness to capture them.
+        exception = None
+        def after_callback():
+            nonlocal exception
+            try:
+                interval_multiplier = next(coroutine) or 1
+            except StopIteration:
+                root.quit()
+            except Exception as exc:
+                exception = exc
+                root.quit()
+            else:
+                root.after(100 * interval_multiplier, after_callback)
+        root.after(0, after_callback)
+        root.mainloop()
+
+        if exception:
+            raise exception
+
+    def test_initial_state(self):
+        sidebar_lines = self.get_sidebar_lines()
+        self.assertEqual(
+            sidebar_lines,
+            ['   '] * (len(sidebar_lines) - 1) + ['>>>'],
+        )
+        self.assert_sidebar_lines_synced()
+
+    @test_coroutine
+    def test_single_empty_input(self):
+        self.do_input('\n')
+        yield 0
+        self.assert_sidebar_lines_end_with(['>>>', '>>>'])
+
+    @test_coroutine
+    def test_single_line_command(self):
+        self.do_input('1\n')
+        yield 2
+        self.assert_sidebar_lines_end_with(['>>>', '   ', '>>>'])
+
+    @test_coroutine
+    def test_multi_line_command(self):
+        # Block statements are not indented because IDLE auto-indents.
+        self.do_input(dedent('''\
+            if True:
+            print(1)
+
+            '''))
+        yield 2
+        self.assert_sidebar_lines_end_with([
+            '>>>',
+            '...',
+            '...',
+            '...',
+            '   ',
+            '>>>',
+        ])
+
+    @test_coroutine
+    def test_single_long_line_wraps(self):
+        self.do_input('1' * 200 + '\n')
+        yield 2
+        self.assert_sidebar_lines_end_with(['>>>', '   ', '>>>'])
+        self.assert_sidebar_lines_synced()
+
+    @test_coroutine
+    def test_squeeze_single_line_command(self):
+        shell = self.shell
+        text = shell.text
+
+        self.do_input('1\n')
+        yield 2
+        self.assert_sidebar_lines_end_with(['>>>', '   ', '>>>'])
+
+        line = int(shell.text.index('insert -1line').split('.')[0])
+        text.mark_set('insert', f"{line}.0")
```

Comment:
```suggestion
        text.mark_set('insert', "insert -1line linestart")
```
This is the only place `line` is used (that I can see) so we could replace with Tk modifiers.

Reply:
Done